Add blog post about advanced usage of specfile (#987)
packit/packit.dev: posts/specfile-advanced/index.md

---
title: Advanced usage of the specfile library
date: 2025-01-09
authors: nforro
tags:
  - specfile
---

# Advanced usage of the specfile library

Are you already familiar with the basic concepts of the _specfile_ library?
Then welcome to a collection of some slightly more advanced topics that you could find useful.

> Otherwise you can take a look at [Introduction to specfile library](posts/specfile-introduction).

Reviewer: _Otherwise_ feels a bit weird here…

1. Maybe you could move it out to a tip admonition below the `<!--truncate-->`
2. Or word it something like:

   ```suggestion
   If you're not that familiar and would like to start with something easier, you can take a look at [Introduction to specfile library](posts/specfile-introduction).
   ```

<!--truncate-->

## Forced parsing

> Imagine you want to make changes to a spec file that requires some sources

Reviewer: #off-topic is there some official glossary with the spelling? :smile: I've seen

* spec file
* specfile
* spec-file

Still not sure which one is _the right one_. :D

to be present at the time of parsing, such as sources referenced from shell

> expansions, e.g. `%(date +%F -r %{SOURCE0})`, or includes some external files
> using the `%include` directive, or loads external macrofiles with the `%load` macro,

Reviewer: ```suggestion
expansions, e.g. `%(date +%F -r %{SOURCE0})`, includes some external files
using the `%include` directive, or loads external macrofiles with the `%load` macro,
```

Not sure about that _e.g._ with all the _or_s

Author: > Not sure about that _e.g._ with all the _or_s

How else would you write it? I wanted to give an example because _sources referenced from shell
expansions_ is a bit vague for most people, I think.

Author: Or do you think it's ok now with the removed or? :slightly_smiling_face:

Reviewer: It's just because of the commas and an example among the enumeration of the possibilities of what user can do. :smile:

Author: Maybe this would be better?

```md
Imagine you want to make changes to a spec file that requires some sources
to be present at the time of parsing such as sources referenced from shell
expansions (e.g. `%(date +%F -r %{SOURCE0})`), includes some external files
using the `%include` directive, or loads external macrofiles with the `%load` macro,
but all you have is the spec file itself.
```

Reviewer: Definitely <3

%global patch_ver 0

Name:       %{pkg_name}
Version:    %{maj_ver}.%{min_ver}.%{patch_ver}%{?rc_ver:~rc%{rc_ver}}
```

It will even uncomment a commented-out macro definition if needed:

```python
spec = Specfile("example.spec")

spec.update_tag("Version", "12.1.1~rc1")
```

This will result in:

```specfile
%global rc_ver 1
%global maj_ver 12
%global min_ver 1
%global patch_ver 1

Name:       %{pkg_name}
Version:    %{maj_ver}.%{min_ver}.%{patch_ver}%{?rc_ver:~rc%{rc_ver}}
```

However, if you want to go back to a regular version, you have to comment out the relevant
macro definition explicitly:

```python
from specfile import Specfile
from specfile.macro_definitions import CommentOutStyle

spec = Specfile("example.spec")

with spec.macro_definitions() as macro_definitions:
    # use the original comment-out style
    macro_definitions.rc_ver.comment_out_style = CommentOutStyle.HASH
    macro_definitions.rc_ver.commented_out = True

spec.update_tag("Version", "12.1.1")
```

This will result in:

```specfile
#global rc_ver 1
%global maj_ver 12
%global min_ver 1
%global patch_ver 1

Name:       %{pkg_name}
Version:    %{maj_ver}.%{min_ver}.%{patch_ver}%{?rc_ver:~rc%{rc_ver}}
```

> The algorithm isn't perfect and won't work in all situations, but even if it's not possible to update

Reviewer 2: Do you think you can describe how the implementation works so people can understand how you match the macros?

Author: I can try, but it would be probably worth a separate blog post :slightly_smiling_face:

Author: I tried to describe it in https://github.com/packit/packit.dev/pull/987/commits/f594540f2d6667eaf8a653e92003e584e354fbc3. WDYT?

Author: Let me merge it as it is, we can always do adjustments later.

some macro definitions it is guaranteed that `expanded_value` of a tag after calling `update_tag()`
will be equal to the given value.

> There is also more general method `Specfile.update_value()` that allows to you update any string

Reviewer 2: ```suggestion
There is also a more general method `Specfile.update_value()` that allows you to update any string
```

set to `True` in order to attempt to parse the spec file with missing external files.
If that succeeds and some required external files were indeed missing during parsing,
the `Specfile.tainted` property is set to `True` to indicate that certain assumptions
could be broken - for example a condition that relies on a macro defined in an external
macrofile could be evaluated incorrectly.

## Defining and undefining macros

You can define or undefine any RPM macros before parsing a spec file with the `macros`
argument to the `Specfile` constructor. The argument takes a list of 2-string tuples
where the first element is macro name and the second element is macro value.
If the second element is `None`, the macro will be undefined. This can come in handy
when you need to undefine a system macro such as `fedora`.

```python
from specfile import Specfile

spec = Specfile(
    "example.spec",
    macros=[
        ("fedora", None),
        ("epel", "9"),
    ],
)
```

It is also possible to pass macro definitions in the same format as the `extra_macros`
argument to the `Specfile.expand()` method in order to temporarily define or undefine macros
before expression expansion.

> For example the following code snippet will print expanded value of the `Release` tag

Reviewer 3: ```suggestion
For example, the following code snippet will print the expanded value of the `Release` tag
```

spec = Specfile("example.spec")

with spec.macro_definitions() as macro_definitions:
    # use the original comment-out style
    macro_definitions.rc_ver.comment_out_style = CommentOutStyle.HASH
    macro_definitions.rc_ver.commented_out = True

spec.update_tag("Version", "12.1.1")
```

This will result in:

```specfile
#global rc_ver 1
%global maj_ver 12
%global min_ver 1
%global patch_ver 1

Name:       %{pkg_name}
Version:    %{maj_ver}.%{min_ver}.%{patch_ver}%{?rc_ver:~rc%{rc_ver}}
```

The algorithm isn't perfect and won't work in all situations, but even if it's not possible to update
some macro definitions it is guaranteed that `expanded_value` of a tag after calling `update_tag()`
will be equal to the given value.

There is also a more general method `Specfile.update_value()` that allows you to update any string
within a spec file in the same manner, trying to preserve macro substitutions and update relevant
macro definitions (or even tag values in case of e.g. `%{version}`) if possible. It is up to you
to replace the original string with the result (that can but doesn't have to be exactly the same).

## More info and links

> Are you interested in more details, trying the library out or even contributing?

Reviewer 3: 👌
